In [1]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns

In [2]:
%matplotlib inline
pd.set_option('display.max_columns', None)

In [3]:
# import excel file 20141104_PrecinctTotals.xlsx
tn_2014=pd.read_excel('../data/2014_election/20141104_PrecinctTotals.xlsx',
                      usecols=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE','RNAME1','PARTY1','PVTALLY1',
                             'RNAME2','PARTY2','PVTALLY2','RNAME3','PARTY3','PVTALLY3','RNAME4','PARTY4','PVTALLY4',
                             'RNAME5','PARTY5','PVTALLY5','RNAME6','PARTY6','PVTALLY6','RNAME7','PARTY7','PVTALLY7',
                             'RNAME8','PARTY8','PVTALLY8','RNAME9','PARTY9','PVTALLY9','RNAME10','PARTY10','PVTALLY10'])
tn_2014

COUNTY       PRECINCT  \
0      Anderson  Andersonville   
1      Anderson  Andersonville   
2      Anderson  Andersonville   
3      Anderson  Andersonville   
4      Anderson  Andersonville   
...         ...            ...   
19981    Wilson    Provisional   
19982    Wilson    Provisional   
19983    Wilson    Provisional   
19984    Wilson    Provisional   
19985    Wilson    Provisional   

                                              OFFICENAME   ELECTDATE  \
0                                               Governor  2014-11-04   
1                             Constitutional Amendment 1  2014-11-04   
2                             Constitutional Amendment 2  2014-11-04   
3                             Constitutional Amendment 3  2014-11-04   
4                             Constitutional Amendment 4  2014-11-04   
...                                                  ...         ...   
19981                               United States Senate  2014-11-04   
19982  United States House of Representatives District 6  2014-11-04   
19983                       Tennessee Senate District 17  2014-11-04   
19984     Tennessee House of Representatives District 46  2014-11-04   
19985     Tennessee House of Representatives District 57  2014-11-04   

               RNAME1       PARTY1  PVTALLY1                      RNAME2  \
0         Bill Haslam   Republican       681  Charles V. "Charlie" Brown   
1                 Yes          NaN       520                          No   
2                 Yes          NaN       568                          No   
3                 Yes          NaN       554                          No   
4                 Yes          NaN       564                          No   
...               ...          ...       ...                         ...   
19981  Eric Schechter  Independent         0                  Rick Tyler   
19982     Diane Black   Republican        17           Amos Scott Powers   
19983     Mae Beavers   Republican        20                         NaN   
19984     Mark A Pody   Republican         6                Candace Reed   
19985    Susan M Lynn   Republican        12              Jesse McLevain   

            PARTY2  PVTALLY2                       RNAME3        PARTY3  \
0       Democratic      96.0                Shaun Crowell  Constitution   
1              NaN     303.0                          NaN           NaN   
2              NaN     247.0                          NaN           NaN   
3              NaN     221.0                          NaN           NaN   
4              NaN     180.0                          NaN           NaN   
...            ...       ...                          ...           ...   
19981  Independent       0.0  Write-In - DeWayne A. Jones   Independent   
19982   Democratic       7.0                  Mike Winton   Independent   
19983          NaN       NaN                          NaN           NaN   
19984   Democratic       2.0                          NaN           NaN   
19985   Democratic       4.0                          NaN           NaN   

       PVTALLY3                      RNAME4       PARTY4  PVTALLY4  \
0          14.0                 Isa Infante        Green       6.0   
1           NaN                         NaN          NaN       NaN   
2           NaN                         NaN          NaN       NaN   
3           NaN                         NaN          NaN       NaN   
4           NaN                         NaN          NaN       NaN   
...         ...                         ...          ...       ...   
19981       0.0  Write-In - Erin Kent Magee  Independent       0.0   
19982       1.0                         NaN          NaN       NaN   
19983       NaN                         NaN          NaN       NaN   
19984       NaN                         NaN          NaN       NaN   
19985       NaN                         NaN          NaN       NaN   

                    RNAME5       PARTY5  PVTALLY5           RNAME6  \
0      Ste

In [4]:
tn_2014.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19986 entries, 0 to 19985
Data columns (total 34 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   COUNTY      19986 non-null  object 
 1   PRECINCT    19986 non-null  object 
 2   OFFICENAME  19986 non-null  object 
 3   ELECTDATE   19986 non-null  object 
 4   RNAME1      19986 non-null  object 
 5   PARTY1      11710 non-null  object 
 6   PVTALLY1    19986 non-null  int64  
 7   RNAME2      18174 non-null  object 
 8   PARTY2      9898 non-null   object 
 9   PVTALLY2    18174 non-null  float64
 10  RNAME3      8695 non-null   object 
 11  PARTY3      8695 non-null   object 
 12  PVTALLY3    8695 non-null   float64
 13  RNAME4      7374 non-null   object 
 14  PARTY4      7374 non-null   object 
 15  PVTALLY4    7374 non-null   float64
 16  RNAME5      4810 non-null   object 
 17  PARTY5      4810 non-null   object 
 18  PVTALLY5    4810 non-null   float64
 19  RNAME6      4140 non-null

In [5]:
# merge entries in all RNAME columns into new column "name"
tn_2014_namemelt=tn_2014.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['RNAME1','RNAME2','RNAME3','RNAME4','RNAME5','RNAME6','RNAME7','RNAME8','RNAME9','RNAME10'],value_name="name")
# drop column "variable" created by .melt
tn_2014_namemelt=tn_2014_namemelt.drop(['variable'],axis=1)                          
tn_2014_namemelt.head(9405)

COUNTY       PRECINCT                  OFFICENAME   ELECTDATE  \
0     Anderson  Andersonville                    Governor  2014-11-04   
1     Anderson  Andersonville  Constitutional Amendment 1  2014-11-04   
2     Anderson  Andersonville  Constitutional Amendment 2  2014-11-04   
3     Anderson  Andersonville  Constitutional Amendment 3  2014-11-04   
4     Anderson  Andersonville  Constitutional Amendment 4  2014-11-04   
...        ...            ...                         ...         ...   
9400   Johnson       3A Trade                    Governor  2014-11-04   
9401   Johnson       3A Trade  Constitutional Amendment 1  2014-11-04   
9402   Johnson       3A Trade  Constitutional Amendment 2  2014-11-04   
9403   Johnson       3A Trade  Constitutional Amendment 3  2014-11-04   
9404   Johnson       3A Trade  Constitutional Amendment 4  2014-11-04   

             name  
0     Bill Haslam  
1             Yes  
2             Yes  
3             Yes  
4             Yes  
...           ...  
9400  Bill Haslam  
9401          Yes  
9402          Yes  
9403          Yes  
9404          Yes  

[9405 rows x 5 columns]

In [6]:
# merge entries in all PARTY columns into new column "party"
tn_2014_partymelt=tn_2014.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['PARTY1','PARTY2','PARTY3','PARTY4','PARTY5','PARTY6','PARTY7','PARTY8','PARTY9','PARTY10'],value_name="party")
# drop column "variable" created by .melt
tn_2014_partymelt=tn_2014_partymelt.drop(['variable'],axis=1)
tn_2014_partymelt.head(9405)

COUNTY       PRECINCT                  OFFICENAME   ELECTDATE  \
0     Anderson  Andersonville                    Governor  2014-11-04   
1     Anderson  Andersonville  Constitutional Amendment 1  2014-11-04   
2     Anderson  Andersonville  Constitutional Amendment 2  2014-11-04   
3     Anderson  Andersonville  Constitutional Amendment 3  2014-11-04   
4     Anderson  Andersonville  Constitutional Amendment 4  2014-11-04   
...        ...            ...                         ...         ...   
9400   Johnson       3A Trade                    Governor  2014-11-04   
9401   Johnson       3A Trade  Constitutional Amendment 1  2014-11-04   
9402   Johnson       3A Trade  Constitutional Amendment 2  2014-11-04   
9403   Johnson       3A Trade  Constitutional Amendment 3  2014-11-04   
9404   Johnson       3A Trade  Constitutional Amendment 4  2014-11-04   

           party  
0     Republican  
1            NaN  
2            NaN  
3            NaN  
4            NaN  
...          ...  
9400  Republican  
9401         NaN  
9402         NaN  
9403         NaN  
9404         NaN  

[9405 rows x 5 columns]

In [7]:
# merge entries in all PVTALLY columns into new column "vote_count"
tn_2014_pvtallymelt=tn_2014.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['PVTALLY1','PVTALLY2','PVTALLY3','PVTALLY4','PVTALLY5','PVTALLY6','PVTALLY7','PVTALLY8','PVTALLY9','PVTALLY10'],value_name="vote_count")
# drop column "variable" created by .melt
tn_2014_pvtallymelt=tn_2014_pvtallymelt.drop(['variable'],axis=1)
tn_2014_pvtallymelt.head(9405)

COUNTY       PRECINCT                  OFFICENAME   ELECTDATE  \
0     Anderson  Andersonville                    Governor  2014-11-04   
1     Anderson  Andersonville  Constitutional Amendment 1  2014-11-04   
2     Anderson  Andersonville  Constitutional Amendment 2  2014-11-04   
3     Anderson  Andersonville  Constitutional Amendment 3  2014-11-04   
4     Anderson  Andersonville  Constitutional Amendment 4  2014-11-04   
...        ...            ...                         ...         ...   
9400   Johnson       3A Trade                    Governor  2014-11-04   
9401   Johnson       3A Trade  Constitutional Amendment 1  2014-11-04   
9402   Johnson       3A Trade  Constitutional Amendment 2  2014-11-04   
9403   Johnson       3A Trade  Constitutional Amendment 3  2014-11-04   
9404   Johnson       3A Trade  Constitutional Amendment 4  2014-11-04   

      vote_count  
0          681.0  
1          520.0  
2          568.0  
3          554.0  
4          564.0  
...          ...  
9400       148.0  
9401       105.0  
9402       127.0  
9403       117.0  
9404       114.0  

[9405 rows x 5 columns]

In [8]:
# add "party" and "vote_count" columns to tn_2014_namemelt
party=tn_2014_partymelt['party']
pvtally=tn_2014_pvtallymelt['vote_count']
tn_2014_namemelt=tn_2014_namemelt.join(party).join(pvtally)
tn_2014_namemelt.head(3)

COUNTY       PRECINCT                  OFFICENAME   ELECTDATE  \
0  Anderson  Andersonville                    Governor  2014-11-04   
1  Anderson  Andersonville  Constitutional Amendment 1  2014-11-04   
2  Anderson  Andersonville  Constitutional Amendment 2  2014-11-04   

          name       party  vote_count  
0  Bill Haslam  Republican       681.0  
1          Yes         NaN       520.0  
2          Yes         NaN       568.0

In [9]:
tn_2014_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 199860 entries, 0 to 199859
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   COUNTY      199860 non-null  object 
 1   PRECINCT    199860 non-null  object 
 2   OFFICENAME  199860 non-null  object 
 3   ELECTDATE   199860 non-null  object 
 4   name        79739 non-null   object 
 5   party       63187 non-null   object 
 6   vote_count  79739 non-null   float64
dtypes: float64(1), object(6)
memory usage: 10.7+ MB


In [10]:
# lower case column names
tn_2014_namemelt.columns=tn_2014_namemelt.columns.str.lower()
tn_2014_namemelt

county       precinct  \
0       Anderson  Andersonville   
1       Anderson  Andersonville   
2       Anderson  Andersonville   
3       Anderson  Andersonville   
4       Anderson  Andersonville   
...          ...            ...   
199855    Wilson    Provisional   
199856    Wilson    Provisional   
199857    Wilson    Provisional   
199858    Wilson    Provisional   
199859    Wilson    Provisional   

                                               officename   electdate  \
0                                                Governor  2014-11-04   
1                              Constitutional Amendment 1  2014-11-04   
2                              Constitutional Amendment 2  2014-11-04   
3                              Constitutional Amendment 3  2014-11-04   
4                              Constitutional Amendment 4  2014-11-04   
...                                                   ...         ...   
199855                               United States Senate  2014-11-04   
199856  United States House of Representatives District 6  2014-11-04   
199857                       Tennessee Senate District 17  2014-11-04   
199858     Tennessee House of Representatives District 46  2014-11-04   
199859     Tennessee House of Representatives District 57  2014-11-04   

               name       party  vote_count  
0       Bill Haslam  Republican       681.0  
1               Yes         NaN       520.0  
2               Yes         NaN       568.0  
3               Yes         NaN       554.0  
4               Yes         NaN       564.0  
...             ...         ...         ...  
199855          NaN         NaN         NaN  
199856          NaN         NaN         NaN  
199857          NaN         NaN         NaN  
199858          NaN         NaN         NaN  
199859          NaN         NaN         NaN  

[199860 rows x 7 columns]

In [11]:
# change null values in "party" column to "no party"
tn_2014_namemelt['party'].fillna("No party", inplace = True)
tn_2014_namemelt

county       precinct  \
0       Anderson  Andersonville   
1       Anderson  Andersonville   
2       Anderson  Andersonville   
3       Anderson  Andersonville   
4       Anderson  Andersonville   
...          ...            ...   
199855    Wilson    Provisional   
199856    Wilson    Provisional   
199857    Wilson    Provisional   
199858    Wilson    Provisional   
199859    Wilson    Provisional   

                                               officename   electdate  \
0                                                Governor  2014-11-04   
1                              Constitutional Amendment 1  2014-11-04   
2                              Constitutional Amendment 2  2014-11-04   
3                              Constitutional Amendment 3  2014-11-04   
4                              Constitutional Amendment 4  2014-11-04   
...                                                   ...         ...   
199855                               United States Senate  2014-11-04   
199856  United States House of Representatives District 6  2014-11-04   
199857                       Tennessee Senate District 17  2014-11-04   
199858     Tennessee House of Representatives District 46  2014-11-04   
199859     Tennessee House of Representatives District 57  2014-11-04   

               name       party  vote_count  
0       Bill Haslam  Republican       681.0  
1               Yes    No party       520.0  
2               Yes    No party       568.0  
3               Yes    No party       554.0  
4               Yes    No party       564.0  
...             ...         ...         ...  
199855          NaN    No party         NaN  
199856          NaN    No party         NaN  
199857          NaN    No party         NaN  
199858          NaN    No party         NaN  
199859          NaN    No party         NaN  

[199860 rows x 7 columns]

In [13]:
tn_2014_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 199860 entries, 0 to 199859
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   county      199860 non-null  object 
 1   precinct    199860 non-null  object 
 2   officename  199860 non-null  object 
 3   electdate   199860 non-null  object 
 4   name        79739 non-null   object 
 5   party       199860 non-null  object 
 6   vote_count  79739 non-null   float64
dtypes: float64(1), object(6)
memory usage: 10.7+ MB


In [14]:
# change null values in "vote_count" column to "0"
tn_2014_namemelt['vote_count'].fillna(0)

0         681.0
1         520.0
2         568.0
3         554.0
4         564.0
          ...  
199855      0.0
199856      0.0
199857      0.0
199858      0.0
199859      0.0
Name: vote_count, Length: 199860, dtype: float64

In [15]:
tn_2014_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 199860 entries, 0 to 199859
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   county      199860 non-null  object 
 1   precinct    199860 non-null  object 
 2   officename  199860 non-null  object 
 3   electdate   199860 non-null  object 
 4   name        79739 non-null   object 
 5   party       199860 non-null  object 
 6   vote_count  79739 non-null   float64
dtypes: float64(1), object(6)
memory usage: 10.7+ MB


### 79729 non-null names and 79739 non-null vote counts
### Do the nulls in "name" correlate one to one with "no party"?

In [16]:
tn_2014_namemelt[tn_2014_namemelt['name'].isna()]

county       precinct  \
19994   Anderson  Andersonville   
20004   Anderson     Briceville   
20005   Anderson     Briceville   
20014   Anderson       Bull Run   
20015   Anderson       Bull Run   
...          ...            ...   
199855    Wilson    Provisional   
199856    Wilson    Provisional   
199857    Wilson    Provisional   
199858    Wilson    Provisional   
199859    Wilson    Provisional   

                                               officename   electdate name  \
19994                         Tennessee Senate District 5  2014-11-04  NaN   
20004                         Tennessee Senate District 5  2014-11-04  NaN   
20005      Tennessee House of Representatives District 33  2014-11-04  NaN   
20014                         Tennessee Senate District 5  2014-11-04  NaN   
20015      Tennessee House of Representatives District 33  2014-11-04  NaN   
...                                                   ...         ...  ...   
199855                               United States Senate  2014-11-04  NaN   
199856  United States House of Representatives District 6  2014-11-04  NaN   
199857                       Tennessee Senate District 17  2014-11-04  NaN   
199858     Tennessee House of Representatives District 46  2014-11-04  NaN   
199859     Tennessee House of Representatives District 57  2014-11-04  NaN   

           party  vote_count  
19994   No party         NaN  
20004   No party         NaN  
20005   No party         NaN  
20014   No party         NaN  
20015   No party         NaN  
...          ...         ...  
199855  No party         NaN  
199856  No party         NaN  
199857  No party         NaN  
199858  No party         NaN  
199859  No party         NaN  

[120121 rows x 7 columns]

In [18]:
tn_2014_namemelt[tn_2014_namemelt['name'].isna()].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 120121 entries, 19994 to 199859
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   county      120121 non-null  object 
 1   precinct    120121 non-null  object 
 2   officename  120121 non-null  object 
 3   electdate   120121 non-null  object 
 4   name        0 non-null       object 
 5   party       120121 non-null  object 
 6   vote_count  0 non-null       float64
dtypes: float64(1), object(6)
memory usage: 7.3+ MB


### Looks safe to drop rows with no votes tallied

In [19]:
# drop rows with no name and vote count
tn_2014_final=tn_2014_namemelt.dropna()
tn_2014_final

county       precinct                  officename   electdate  \
0       Anderson  Andersonville                    Governor  2014-11-04   
1       Anderson  Andersonville  Constitutional Amendment 1  2014-11-04   
2       Anderson  Andersonville  Constitutional Amendment 2  2014-11-04   
3       Anderson  Andersonville  Constitutional Amendment 3  2014-11-04   
4       Anderson  Andersonville  Constitutional Amendment 4  2014-11-04   
...          ...            ...                         ...         ...   
199832    Wilson           25-2        United States Senate  2014-11-04   
199838    Wilson       Absentee                    Governor  2014-11-04   
199843    Wilson       Absentee        United States Senate  2014-11-04   
199849    Wilson    Provisional                    Governor  2014-11-04   
199854    Wilson    Provisional        United States Senate  2014-11-04   

                           name        party  vote_count  
0                   Bill Haslam   Republican       681.0  
1                           Yes     No party       520.0  
2                           Yes     No party       568.0  
3                           Yes     No party       554.0  
4                           Yes     No party       564.0  
...                         ...          ...         ...  
199832               C. Salekin  Independent         1.0  
199838  Write-In - Mark Pearson  Independent         0.0  
199843               C. Salekin  Independent         1.0  
199849  Write-In - Mark Pearson  Independent         0.0  
199854               C. Salekin  Independent         0.0  

[79739 rows x 7 columns]

In [20]:
tn_2014_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 79739 entries, 0 to 199854
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   county      79739 non-null  object 
 1   precinct    79739 non-null  object 
 2   officename  79739 non-null  object 
 3   electdate   79739 non-null  object 
 4   name        79739 non-null  object 
 5   party       79739 non-null  object 
 6   vote_count  79739 non-null  float64
dtypes: float64(1), object(6)
memory usage: 4.9+ MB


In [21]:
# convert vote_count data from float64 to int 32 to to drop ".0"
tn_2014_final=tn_2014_final.astype({'vote_count':'int32'})
tn_2014_final

county       precinct                  officename   electdate  \
0       Anderson  Andersonville                    Governor  2014-11-04   
1       Anderson  Andersonville  Constitutional Amendment 1  2014-11-04   
2       Anderson  Andersonville  Constitutional Amendment 2  2014-11-04   
3       Anderson  Andersonville  Constitutional Amendment 3  2014-11-04   
4       Anderson  Andersonville  Constitutional Amendment 4  2014-11-04   
...          ...            ...                         ...         ...   
199832    Wilson           25-2        United States Senate  2014-11-04   
199838    Wilson       Absentee                    Governor  2014-11-04   
199843    Wilson       Absentee        United States Senate  2014-11-04   
199849    Wilson    Provisional                    Governor  2014-11-04   
199854    Wilson    Provisional        United States Senate  2014-11-04   

                           name        party  vote_count  
0                   Bill Haslam   Republican         681  
1                           Yes     No party         520  
2                           Yes     No party         568  
3                           Yes     No party         554  
4                           Yes     No party         564  
...                         ...          ...         ...  
199832               C. Salekin  Independent           1  
199838  Write-In - Mark Pearson  Independent           0  
199843               C. Salekin  Independent           1  
199849  Write-In - Mark Pearson  Independent           0  
199854               C. Salekin  Independent           0  

[79739 rows x 7 columns]

In [22]:
tn_2014_final.dtypes

county        object
precinct      object
officename    object
electdate     object
name          object
party         object
vote_count     int32
dtype: object

In [23]:
tn_2014_final['officename'].value_counts()

United States Senate                              28966
Governor                                          20710
Constitutional Amendment 4                         4138
Constitutional Amendment 1                         4138
Constitutional Amendment 3                         4138
                                                  ...  
Tennessee House of Representatives District 16       10
Tennessee House of Representatives District 18       10
Tennessee House of Representatives District 24       10
Tennessee House of Representatives District 81        9
Tennessee House of Representatives District 14        7
Name: officename, Length: 132, dtype: int64

In [24]:
tn_2014_final['officename'].unique()

array(['Governor', 'Constitutional Amendment 1',
       'Constitutional Amendment 2', 'Constitutional Amendment 3',
       'Constitutional Amendment 4', 'United States Senate',
       'United States House of Representatives District 3',
       'Tennessee Senate District 5',
       'Tennessee House of Representatives District 36',
       'Tennessee House of Representatives District 33',
       'United States House of Representatives District 4',
       'Tennessee House of Representatives District 62',
       'United States House of Representatives District 7',
       'Tennessee House of Representatives District 75',
       'United States House of Representatives District 8',
       'Tennessee Senate District 15',
       'Tennessee House of Representatives District 31',
       'United States House of Representatives District 2',
       'Tennessee House of Representatives District 8',
       'Tennessee House of Representatives District 20',
       'Tennessee House of Representatives Distr

In [25]:
tn_2014_final['vote_count'].value_counts()

0       18467
1        5332
2        3863
3        2830
4        2120
        ...  
1692        1
1628        1
1564        1
1372        1
1759        1
Name: vote_count, Length: 1716, dtype: int64

### 18467 rows have a vote count of 0. Drop those rows

In [26]:
tn_2014_final=tn_2014_final[tn_2014_final['vote_count'] != 0]
tn_2014_final=tn_2014_final.reset_index(drop=True)
tn_2014_final.head()

county       precinct                  officename   electdate  \
0  Anderson  Andersonville                    Governor  2014-11-04   
1  Anderson  Andersonville  Constitutional Amendment 1  2014-11-04   
2  Anderson  Andersonville  Constitutional Amendment 2  2014-11-04   
3  Anderson  Andersonville  Constitutional Amendment 3  2014-11-04   
4  Anderson  Andersonville  Constitutional Amendment 4  2014-11-04   

          name       party  vote_count  
0  Bill Haslam  Republican         681  
1          Yes    No party         520  
2          Yes    No party         568  
3          Yes    No party         554  
4          Yes    No party         564

In [27]:
# confirm no 0's in vote count
tn_2014_final['vote_count'].value_counts()

1       5332
2       3863
3       2830
4       2120
5       1782
        ... 
2024       1
2868       1
1918       1
1854       1
2047       1
Name: vote_count, Length: 1715, dtype: int64

In [29]:
# create data frame tn_2014_uss for senatorial election
tn_2014_uss_int=tn_2014_final[tn_2014_final['officename'].str.match('United States Senate')]
tn_2014_uss_int=tn_2014_uss_int.reset_index(drop=True)
tn_2014_uss_int

county       precinct            officename   electdate  \
0      Anderson  Andersonville  United States Senate  2014-11-04   
1      Anderson  Andersonville  United States Senate  2014-11-04   
2      Anderson     Briceville  United States Senate  2014-11-04   
3      Anderson       Bull Run  United States Senate  2014-11-04   
4      Anderson       Bull Run  United States Senate  2014-11-04   
...         ...            ...                   ...         ...   
18696    Wilson           14-1  United States Senate  2014-11-04   
18697    Wilson           23-1  United States Senate  2014-11-04   
18698    Wilson           25-1  United States Senate  2014-11-04   
18699    Wilson           25-2  United States Senate  2014-11-04   
18700    Wilson       Absentee  United States Senate  2014-11-04   

                  name        party  vote_count  
0      Lamar Alexander   Republican         603  
1       Eric Schechter  Independent           1  
2      Lamar Alexander   Republican         101  
3      Lamar Alexander   Republican         844  
4       Eric Schechter  Independent           1  
...                ...          ...         ...  
18696       C. Salekin  Independent           1  
18697       C. Salekin  Independent           1  
18698       C. Salekin  Independent           1  
18699       C. Salekin  Independent           1  
18700       C. Salekin  Independent           1  

[18701 rows x 7 columns]

In [30]:
tn_2014_uss_int.head(20)

county        precinct            officename   electdate  \
0   Anderson   Andersonville  United States Senate  2014-11-04   
1   Anderson   Andersonville  United States Senate  2014-11-04   
2   Anderson      Briceville  United States Senate  2014-11-04   
3   Anderson        Bull Run  United States Senate  2014-11-04   
4   Anderson        Bull Run  United States Senate  2014-11-04   
5   Anderson         Clinton  United States Senate  2014-11-04   
6   Anderson         Clinton  United States Senate  2014-11-04   
7   Anderson    Clinton High  United States Senate  2014-11-04   
8   Anderson    Clinton High  United States Senate  2014-11-04   
9   Anderson         Claxton  United States Senate  2014-11-04   
10  Anderson         Claxton  United States Senate  2014-11-04   
11  Anderson  Clinton Middle  United States Senate  2014-11-04   
12  Anderson    Dutch Valley  United States Senate  2014-11-04   
13  Anderson    Emory Valley  United States Senate  2014-11-04   
14  Anderson    Emory Valley  United States Senate  2014-11-04   
15  Anderson        Fairview  United States Senate  2014-11-04   
16  Anderson     Glen Alpine  United States Senate  2014-11-04   
17  Anderson        Glenwood  United States Senate  2014-11-04   
18  Anderson   Hendrix Creek  United States Senate  2014-11-04   
19  Anderson   Hendrix Creek  United States Senate  2014-11-04   

               name        party  vote_count  
0   Lamar Alexander   Republican         603  
1    Eric Schechter  Independent           1  
2   Lamar Alexander   Republican         101  
3   Lamar Alexander   Republican         844  
4    Eric Schechter  Independent           1  
5   Lamar Alexander   Republican         896  
6    Eric Schechter  Independent           1  
7   Lamar Alexander   Republican         444  
8    Eric Schechter  Independent           2  
9   Lamar Alexander   Republican         677  
10   Eric Schechter  Independent           1  
11  Lamar Alexander   Republican         202  
12  Lamar Alexander   Republican         156  
13  Lamar Alexander   Republican         942  
14   Eric Schechter  Independent           1  
15  Lamar Alexander   Republican         257  
16  Lamar Alexander   Republican         562  
17  Lamar Alexander   Republican         487  
18  Lamar Alexander   Republican         763  
19   Eric Schechter  Independent           2

### Add column that has total vote count by count for use in calculating county vote percentages

In [31]:
# group county precincts into county per candidate, in descending order of votes within county
tn_2014_uss_int=tn_2014_uss_int.groupby(['electdate','officename','county','name','party'], as_index=False)['vote_count'].sum()
tn_2014_uss=tn_2014_uss_int.sort_values(by=['county','vote_count'],ascending=False).reset_index(drop=True)
tn_2014_uss.head(30)

electdate            officename      county                        name  \
0   2014-11-04  United States Senate      Wilson             Lamar Alexander   
1   2014-11-04  United States Senate      Wilson                 Gordon Ball   
2   2014-11-04  United States Senate      Wilson                 Joe Wilmoth   
3   2014-11-04  United States Senate      Wilson                  Danny Page   
4   2014-11-04  United States Senate      Wilson            Tom Emerson, Jr.   
5   2014-11-04  United States Senate      Wilson             Martin Pleasant   
6   2014-11-04  United States Senate      Wilson                Joshua James   
7   2014-11-04  United States Senate      Wilson                  Rick Tyler   
8   2014-11-04  United States Senate      Wilson          Edmund L. Gauthier   
9   2014-11-04  United States Senate      Wilson     Bartholomew J. Phillips   
10  2014-11-04  United States Senate      Wilson              Eric Schechter   
11  2014-11-04  United States Senate      Wilson                  C. Salekin   
12  2014-11-04  United States Senate  Williamson             Lamar Alexander   
13  2014-11-04  United States Senate  Williamson                 Gordon Ball   
14  2014-11-04  United States Senate  Williamson                 Joe Wilmoth   
15  2014-11-04  United States Senate  Williamson             Martin Pleasant   
16  2014-11-04  United States Senate  Williamson            Tom Emerson, Jr.   
17  2014-11-04  United States Senate  Williamson                  Danny Page   
18  2014-11-04  United States Senate  Williamson                Joshua James   
19  2014-11-04  United States Senate  Williamson                  Rick Tyler   
20  2014-11-04  United States Senate  Williamson     Bartholomew J. Phillips   
21  2014-11-04  United States Senate  Williamson          Edmund L. Gauthier   
22  2014-11-04  United States Senate  Williamson              Eric Schechter   
23  2014-11-04  United States Senate  Williamson                  C. Salekin   
24  2014-11-04  United States Senate  Williamson  Write-In - Erin Kent Magee   
25  2014-11-04  United States Senate       White             Lamar Alexander   
26  2014-11-04  United States Senate       White                 Gordon Ball   
27  2014-11-04  United States Senate       White                 Joe Wilmoth   
28  2014-11-04  United States Senate       White            Tom Emerson, Jr.   
29  2014-11-04  United States Senate       White                  Danny Page   

           party  vote_count  
0     Republican       20260  
1     Democratic        7903  
2   Constitution         911  
3    Independent         363  
4    Independent         272  
5          Green         210  
6    Independent         148  
7    Independent         130  
8    Independent          60  
9    Independent          53  
10   Independent          38  
11   Independent          16  
12    Republican       42501  
13    Democratic       12660  
14  Constitution        1816  
15         Green         527  
16   Independent         518  
17   Independent         320  
18   Independent         230  
19   Independent         178  
20   Independent          91  
21   Independent          82  
22   Independent          67  
23   Independent          28  
24   Independent           1  
25    Republican        3721  
26    Democratic        1678  
27  Constitution         309  
28   Independent          57  
29   Independent          53

In [33]:
# create new dataframe "county_vote" to use to find sum of votes per county
county_vote=tn_2014_uss.loc[0:,['county','vote_count']]
county_vote

county  vote_count
0       Wilson       20260
1       Wilson        7903
2       Wilson         911
3       Wilson         363
4       Wilson         272
...        ...         ...
1128  Anderson          64
1129  Anderson          47
1130  Anderson          36
1131  Anderson          22
1132  Anderson          14

[1133 rows x 2 columns]

In [34]:
#sum county votes
county_vote_sum=county_vote.groupby(['county'])['vote_count'].sum()
county_vote_sum

county
Anderson      17994
Bedford        7564
Benton         3739
Bledsoe        2609
Blount        29868
              ...  
Wayne          2730
Weakley        7502
White          5964
Williamson    59019
Wilson        30364
Name: vote_count, Length: 95, dtype: int32

In [35]:
# use "county_vote_sum" to add column "sum_county_vote" to tn_2018_gov correlating with value in each row
tn_2014_uss['sum_county_vote']=tn_2014_uss['county'].map(county_vote_sum)
tn_2014_uss.head(20)

electdate            officename      county                     name  \
0   2014-11-04  United States Senate      Wilson          Lamar Alexander   
1   2014-11-04  United States Senate      Wilson              Gordon Ball   
2   2014-11-04  United States Senate      Wilson              Joe Wilmoth   
3   2014-11-04  United States Senate      Wilson               Danny Page   
4   2014-11-04  United States Senate      Wilson         Tom Emerson, Jr.   
5   2014-11-04  United States Senate      Wilson          Martin Pleasant   
6   2014-11-04  United States Senate      Wilson             Joshua James   
7   2014-11-04  United States Senate      Wilson               Rick Tyler   
8   2014-11-04  United States Senate      Wilson       Edmund L. Gauthier   
9   2014-11-04  United States Senate      Wilson  Bartholomew J. Phillips   
10  2014-11-04  United States Senate      Wilson           Eric Schechter   
11  2014-11-04  United States Senate      Wilson               C. Salekin   
12  2014-11-04  United States Senate  Williamson          Lamar Alexander   
13  2014-11-04  United States Senate  Williamson              Gordon Ball   
14  2014-11-04  United States Senate  Williamson              Joe Wilmoth   
15  2014-11-04  United States Senate  Williamson          Martin Pleasant   
16  2014-11-04  United States Senate  Williamson         Tom Emerson, Jr.   
17  2014-11-04  United States Senate  Williamson               Danny Page   
18  2014-11-04  United States Senate  Williamson             Joshua James   
19  2014-11-04  United States Senate  Williamson               Rick Tyler   

           party  vote_count  sum_county_vote  
0     Republican       20260            30364  
1     Democratic        7903            30364  
2   Constitution         911            30364  
3    Independent         363            30364  
4    Independent         272            30364  
5          Green         210            30364  
6    Independent         148            30364  
7    Independent         130            30364  
8    Independent          60            30364  
9    Independent          53            30364  
10   Independent          38            30364  
11   Independent          16            30364  
12    Republican       42501            59019  
13    Democratic       12660            59019  
14  Constitution        1816            59019  
15         Green         527            59019  
16   Independent         518            59019  
17   Independent         320            59019  
18   Independent         230            59019  
19   Independent         178            59019

In [36]:
# create "pct_county_vote" column for tn_2016_pres
tn_2014_uss['pct_county_vote']=round(tn_2014_uss['vote_count']/tn_2014_uss['sum_county_vote']*100,2)
tn_2014_uss

electdate            officename    county                     name  \
0     2014-11-04  United States Senate    Wilson          Lamar Alexander   
1     2014-11-04  United States Senate    Wilson              Gordon Ball   
2     2014-11-04  United States Senate    Wilson              Joe Wilmoth   
3     2014-11-04  United States Senate    Wilson               Danny Page   
4     2014-11-04  United States Senate    Wilson         Tom Emerson, Jr.   
...          ...                   ...       ...                      ...   
1128  2014-11-04  United States Senate  Anderson               Danny Page   
1129  2014-11-04  United States Senate  Anderson       Edmund L. Gauthier   
1130  2014-11-04  United States Senate  Anderson  Bartholomew J. Phillips   
1131  2014-11-04  United States Senate  Anderson           Eric Schechter   
1132  2014-11-04  United States Senate  Anderson               C. Salekin   

             party  vote_count  sum_county_vote  pct_county_vote  
0       Republican       20260            30364            66.72  
1       Democratic        7903            30364            26.03  
2     Constitution         911            30364             3.00  
3      Independent         363            30364             1.20  
4      Independent         272            30364             0.90  
...            ...         ...              ...              ...  
1128   Independent          64            17994             0.36  
1129   Independent          47            17994             0.26  
1130   Independent          36            17994             0.20  
1131   Independent          22            17994             0.12  
1132   Independent          14            17994             0.08  

[1133 rows x 8 columns]

In [38]:
# how many parties?
tn_2014_uss['party'].value_counts()

Independent     753
Democratic       95
Green            95
Republican       95
Constitution     95
Name: party, dtype: int64

### Create dataframe "tn_2014_uss_pct_county"  with percent values per party per county

In [39]:
# create new dataframe with reduced columns values only
tn_2014_uss_pct_county=tn_2014_uss.loc[0:,['electdate','officename','county','party','pct_county_vote']]
tn_2014_uss_pct_county

electdate            officename    county         party  \
0     2014-11-04  United States Senate    Wilson    Republican   
1     2014-11-04  United States Senate    Wilson    Democratic   
2     2014-11-04  United States Senate    Wilson  Constitution   
3     2014-11-04  United States Senate    Wilson   Independent   
4     2014-11-04  United States Senate    Wilson   Independent   
...          ...                   ...       ...           ...   
1128  2014-11-04  United States Senate  Anderson   Independent   
1129  2014-11-04  United States Senate  Anderson   Independent   
1130  2014-11-04  United States Senate  Anderson   Independent   
1131  2014-11-04  United States Senate  Anderson   Independent   
1132  2014-11-04  United States Senate  Anderson   Independent   

      pct_county_vote  
0               66.72  
1               26.03  
2                3.00  
3                1.20  
4                0.90  
...               ...  
1128             0.36  
1129             0.26  
1130             0.20  
1131             0.12  
1132             0.08  

[1133 rows x 5 columns]

In [42]:
# create new dataframe with Republican values only
tn_2014_uss_rep_pct=tn_2014_uss_pct_county[tn_2014_uss_pct_county['party']== "Republican"].reset_index(drop=True)
tn_2014_uss_rep_pct

electdate            officename      county       party  pct_county_vote
0   2014-11-04  United States Senate      Wilson  Republican            66.72
1   2014-11-04  United States Senate  Williamson  Republican            72.01
2   2014-11-04  United States Senate       White  Republican            62.39
3   2014-11-04  United States Senate     Weakley  Republican            67.54
4   2014-11-04  United States Senate       Wayne  Republican            73.88
..         ...                   ...         ...         ...              ...
90  2014-11-04  United States Senate      Blount  Republican            68.84
91  2014-11-04  United States Senate     Bledsoe  Republican            69.45
92  2014-11-04  United States Senate      Benton  Republican            58.89
93  2014-11-04  United States Senate     Bedford  Republican            66.68
94  2014-11-04  United States Senate    Anderson  Republican            64.53

[95 rows x 5 columns]

In [43]:
# rename 'pct_county_vote' to 'rep_pct' and drop "party" column to prepare for merge 
tn_2014_uss_rep_pct=tn_2014_uss_rep_pct.rename(columns={'pct_county_vote':'Republican'},inplace=False).drop(['party'], axis=1)
tn_2014_uss_rep_pct

electdate            officename      county  Republican
0   2014-11-04  United States Senate      Wilson       66.72
1   2014-11-04  United States Senate  Williamson       72.01
2   2014-11-04  United States Senate       White       62.39
3   2014-11-04  United States Senate     Weakley       67.54
4   2014-11-04  United States Senate       Wayne       73.88
..         ...                   ...         ...         ...
90  2014-11-04  United States Senate      Blount       68.84
91  2014-11-04  United States Senate     Bledsoe       69.45
92  2014-11-04  United States Senate      Benton       58.89
93  2014-11-04  United States Senate     Bedford       66.68
94  2014-11-04  United States Senate    Anderson       64.53

[95 rows x 4 columns]

In [45]:
# create new dataframe with Democratic values only
tn_2014_uss_dem_pct=tn_2014_uss_pct_county[tn_2014_uss_pct_county['party']== "Democratic"].reset_index(drop=True)
tn_2014_uss_dem_pct

electdate            officename      county       party  pct_county_vote
0   2014-11-04  United States Senate      Wilson  Democratic            26.03
1   2014-11-04  United States Senate  Williamson  Democratic            21.45
2   2014-11-04  United States Senate       White  Democratic            28.14
3   2014-11-04  United States Senate     Weakley  Democratic            24.69
4   2014-11-04  United States Senate       Wayne  Democratic            19.60
..         ...                   ...         ...         ...              ...
90  2014-11-04  United States Senate      Blount  Democratic            23.97
91  2014-11-04  United States Senate     Bledsoe  Democratic            24.34
92  2014-11-04  United States Senate      Benton  Democratic            34.58
93  2014-11-04  United States Senate     Bedford  Democratic            26.37
94  2014-11-04  United States Senate    Anderson  Democratic            29.24

[95 rows x 5 columns]

In [46]:
# rename 'pct_county_vote' to 'dem_pct' and drop "party" to prepare for merge 
tn_2014_uss_dem_pct=tn_2014_uss_dem_pct.rename(columns={'pct_county_vote':'Democratic'},inplace=False).drop(['party'], axis=1)
tn_2014_uss_dem_pct

electdate            officename      county  Democratic
0   2014-11-04  United States Senate      Wilson       26.03
1   2014-11-04  United States Senate  Williamson       21.45
2   2014-11-04  United States Senate       White       28.14
3   2014-11-04  United States Senate     Weakley       24.69
4   2014-11-04  United States Senate       Wayne       19.60
..         ...                   ...         ...         ...
90  2014-11-04  United States Senate      Blount       23.97
91  2014-11-04  United States Senate     Bledsoe       24.34
92  2014-11-04  United States Senate      Benton       34.58
93  2014-11-04  United States Senate     Bedford       26.37
94  2014-11-04  United States Senate    Anderson       29.24

[95 rows x 4 columns]

In [47]:
# create new dataframe with non Republican or Democratic values only
tn_2014_uss_other_pct=tn_2014_uss_pct_county[(tn_2014_uss_pct_county['party']!="Republican")&(tn_2014_uss_pct_county['party']!="Democratic")].reset_index(drop=True)
tn_2014_uss_other_pct

electdate            officename    county         party  pct_county_vote
0    2014-11-04  United States Senate    Wilson  Constitution             3.00
1    2014-11-04  United States Senate    Wilson   Independent             1.20
2    2014-11-04  United States Senate    Wilson   Independent             0.90
3    2014-11-04  United States Senate    Wilson         Green             0.69
4    2014-11-04  United States Senate    Wilson   Independent             0.49
..          ...                   ...       ...           ...              ...
938  2014-11-04  United States Senate  Anderson   Independent             0.36
939  2014-11-04  United States Senate  Anderson   Independent             0.26
940  2014-11-04  United States Senate  Anderson   Independent             0.20
941  2014-11-04  United States Senate  Anderson   Independent             0.12
942  2014-11-04  United States Senate  Anderson   Independent             0.08

[943 rows x 5 columns]

In [48]:
# sum by county
tn_2014_uss_other_pct=tn_2014_uss_other_pct.groupby(['county'],as_index=False)['pct_county_vote'].sum()
tn_2014_uss_other_pct

county  pct_county_vote
0     Anderson             6.22
1      Bedford             6.95
2       Benton             6.53
3      Bledsoe             6.20
4       Blount             7.20
..         ...              ...
90       Wayne             6.52
91     Weakley             7.77
92       White             9.47
93  Williamson             6.53
94      Wilson             7.26

[95 rows x 2 columns]

In [49]:
# rename 'pct_county_vote' to 'other_pct' and drop "party" to prepare for merge 
tn_2014_uss_other_pct=tn_2014_uss_other_pct.rename(columns={'pct_county_vote':'Other'})
tn_2014_uss_other_pct

county  Other
0     Anderson   6.22
1      Bedford   6.95
2       Benton   6.53
3      Bledsoe   6.20
4       Blount   7.20
..         ...    ...
90       Wayne   6.52
91     Weakley   7.77
92       White   9.47
93  Williamson   6.53
94      Wilson   7.26

[95 rows x 2 columns]

In [50]:
# join tn_2014_uss_rep, tn_2014_uss_dem_pct and tn_2014_uss_other_pct to form tn_2014_uss_county_pct
dem=tn_2014_uss_dem_pct['Democratic']
other=tn_2014_uss_other_pct['Other']
tn_2014_uss_county_pct=tn_2014_uss_rep_pct.join(dem).join(other)
tn_2014_uss_county_pct

electdate            officename      county  Republican  Democratic  \
0   2014-11-04  United States Senate      Wilson       66.72       26.03   
1   2014-11-04  United States Senate  Williamson       72.01       21.45   
2   2014-11-04  United States Senate       White       62.39       28.14   
3   2014-11-04  United States Senate     Weakley       67.54       24.69   
4   2014-11-04  United States Senate       Wayne       73.88       19.60   
..         ...                   ...         ...         ...         ...   
90  2014-11-04  United States Senate      Blount       68.84       23.97   
91  2014-11-04  United States Senate     Bledsoe       69.45       24.34   
92  2014-11-04  United States Senate      Benton       58.89       34.58   
93  2014-11-04  United States Senate     Bedford       66.68       26.37   
94  2014-11-04  United States Senate    Anderson       64.53       29.24   

    Other  
0    6.22  
1    6.95  
2    6.53  
3    6.20  
4    7.20  
..    ...  
90   6.52  
91   7.77  
92   9.47  
93   6.53  
94   7.26  

[95 rows x 6 columns]

### Add column for Republican margin over Democrat for heat mapping in Tableau

In [51]:
tn_2014_uss_county_pct['Rep_Margin']= tn_2014_uss_county_pct['Republican']-tn_2014_uss_county_pct['Democratic']
tn_2014_uss_county_pct

electdate            officename      county  Republican  Democratic  \
0   2014-11-04  United States Senate      Wilson       66.72       26.03   
1   2014-11-04  United States Senate  Williamson       72.01       21.45   
2   2014-11-04  United States Senate       White       62.39       28.14   
3   2014-11-04  United States Senate     Weakley       67.54       24.69   
4   2014-11-04  United States Senate       Wayne       73.88       19.60   
..         ...                   ...         ...         ...         ...   
90  2014-11-04  United States Senate      Blount       68.84       23.97   
91  2014-11-04  United States Senate     Bledsoe       69.45       24.34   
92  2014-11-04  United States Senate      Benton       58.89       34.58   
93  2014-11-04  United States Senate     Bedford       66.68       26.37   
94  2014-11-04  United States Senate    Anderson       64.53       29.24   

    Other  Rep_Margin  
0    6.22       40.69  
1    6.95       50.56  
2    6.53       34.25  
3    6.20       42.85  
4    7.20       54.28  
..    ...         ...  
90   6.52       44.87  
91   7.77       45.11  
92   9.47       24.31  
93   6.53       40.31  
94   7.26       35.29  

[95 rows x 7 columns]

In [52]:
# change "Dakalb" to "DeKalb"
tn_2014_uss_county_pct['county']=tn_2014_uss_county_pct['county'].str.replace('Dekalb','DeKalb')
tn_2014_uss_county_pct.tail(21)

electdate            officename      county  Republican  Democratic  \
74  2014-11-04  United States Senate      DeKalb       59.61       32.79   
75  2014-11-04  United States Senate     Decatur       63.32       30.26   
76  2014-11-04  United States Senate    Davidson       43.76       50.83   
77  2014-11-04  United States Senate  Cumberland       69.74       22.78   
78  2014-11-04  United States Senate    Crockett       67.97       25.30   
79  2014-11-04  United States Senate      Coffee       65.47       26.82   
80  2014-11-04  United States Senate       Cocke       60.08       34.19   
81  2014-11-04  United States Senate        Clay       61.61       31.86   
82  2014-11-04  United States Senate   Claiborne       69.32       24.88   
83  2014-11-04  United States Senate     Chester       70.74       18.98   
84  2014-11-04  United States Senate    Cheatham       64.32       27.46   
85  2014-11-04  United States Senate      Carter       73.54       18.91   
86  2014-11-04  United States Senate     Carroll       65.61       26.90   
87  2014-11-04  United States Senate      Cannon       62.36       29.08   
88  2014-11-04  United States Senate    Campbell       70.46       24.14   
89  2014-11-04  United States Senate     Bradley       78.87       16.02   
90  2014-11-04  United States Senate      Blount       68.84       23.97   
91  2014-11-04  United States Senate     Bledsoe       69.45       24.34   
92  2014-11-04  United States Senate      Benton       58.89       34.58   
93  2014-11-04  United States Senate     Bedford       66.68       26.37   
94  2014-11-04  United States Senate    Anderson       64.53       29.24   

    Other  Rep_Margin  
74   7.61       26.82  
75   5.36       33.06  
76   5.95       -7.07  
77   6.41       46.96  
78   4.02       42.67  
79   7.18       38.65  
80   6.86       25.89  
81   7.17       29.75  
82   6.85       44.44  
83   7.10       51.76  
84   8.35       36.86  
85   8.01       54.63  
86   6.49       38.71  
87   7.50       33.28  
88   6.74       46.32  
89   7.64       62.85  
90   6.52       44.87  
91   7.77       45.11  
92   9.47       24.31  
93   6.53       40.31  
94   7.26       35.29

In [85]:
#read TN_county_index_shape.csv to add index from TN Counties shape file for joining in Tableau
county_index=pd.read_csv('../data/TN_county_index_shape.csv')
county_index


county  INDEX_NO
0     Anderson        79
1      Bedford        26
2       Benton        41
3      Bledsoe        31
4       Blount        52
..         ...       ...
90       Wayne         6
91     Weakley        56
92       White        67
93  Williamson        44
94      Wilson        65

[95 rows x 2 columns]

In [87]:
# join with tn_2014_uss_county_pct
tn_2014_g_county_pct=tn_2014_g_county_pct.merge(county_index,on="county")
tn_2014_g_county_pct.tail(25)

electdate officename      county  Republican  Democratic  Other  \
70  2014-11-04   Governor    Fentress       73.20       19.05   8.24   
71  2014-11-04   Governor     Fayette       77.14       18.72   4.24   
72  2014-11-04   Governor        Dyer       77.75       15.82   6.84   
73  2014-11-04   Governor     Dickson       69.09       23.24   8.06   
74  2014-11-04   Governor      DeKalb       67.85       23.35   8.09   
75  2014-11-04   Governor     Decatur       71.29       22.25   4.50   
76  2014-11-04   Governor    Davidson       53.50       35.41   4.48   
77  2014-11-04   Governor  Cumberland       79.01       15.48   5.80   
78  2014-11-04   Governor    Crockett       75.84       19.04   4.83   
79  2014-11-04   Governor      Coffee       73.11       18.38   8.68   
80  2014-11-04   Governor       Cocke       78.30       14.39   7.48   
81  2014-11-04   Governor        Clay       66.64       26.16   5.67   
82  2014-11-04   Governor   Claiborne       76.20       17.99   6.53   
83  2014-11-04   Governor     Chester       78.58       15.07   5.08   
84  2014-11-04   Governor    Cheatham       72.63       18.01   8.06   
85  2014-11-04   Governor      Carter       77.72       14.20   8.52   
86  2014-11-04   Governor     Carroll       69.54       23.06   5.75   
87  2014-11-04   Governor      Cannon       71.37       19.71   8.32   
88  2014-11-04   Governor    Campbell       77.42       17.80   8.08   
89  2014-11-04   Governor     Bradley       83.19       12.59   8.01   
90  2014-11-04   Governor      Blount       79.75       14.06   6.85   
91  2014-11-04   Governor     Bledsoe       69.41       24.10   9.28   
92  2014-11-04   Governor      Benton       64.73       26.56   8.02   
93  2014-11-04   Governor     Bedford       75.98       16.49   7.08   
94  2014-11-04   Governor    Anderson       75.27       17.86   6.98   

    Rep_Margin  INDEX_NO  
70       54.15        76  
71       58.42         2  
72       61.93        54  
73       45.85        43  
74       44.50        66  
75       49.04        20  
76       18.09        63  
77       63.53        49  
78       56.80        38  
79       54.73        28  
80       63.91        88  
81       40.48        73  
82       58.21        82  
83       63.51        18  
84       54.62        61  
85       63.52        94  
86       46.48        40  
87       51.66        46  
88       59.62        81  
89       70.60        13  
90       65.69        52  
91       45.31        31  
92       38.17        41  
93       59.49        26  
94       57.41        79

In [53]:
tn_2014_uss_county_pct.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 95 entries, 0 to 94
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   electdate   95 non-null     object 
 1   officename  95 non-null     object 
 2   county      95 non-null     object 
 3   Republican  95 non-null     float64
 4   Democratic  95 non-null     float64
 5   Other       95 non-null     float64
 6   Rep_Margin  95 non-null     float64
dtypes: float64(4), object(3)
memory usage: 5.3+ KB


In [54]:
# covert "electdate" to datetime from object
tn_2014_uss_county_pct['electdate']=pd.to_datetime(tn_2014_uss_county_pct['electdate'],format='%Y-%m-%d')
tn_2014_uss_county_pct.tail(25)

electdate            officename      county  Republican  Democratic  \
70 2014-11-04  United States Senate    Fentress       69.02       22.82   
71 2014-11-04  United States Senate     Fayette       70.28       23.90   
72 2014-11-04  United States Senate        Dyer       69.53       21.49   
73 2014-11-04  United States Senate     Dickson       61.27       30.97   
74 2014-11-04  United States Senate      DeKalb       59.61       32.79   
75 2014-11-04  United States Senate     Decatur       63.32       30.26   
76 2014-11-04  United States Senate    Davidson       43.76       50.83   
77 2014-11-04  United States Senate  Cumberland       69.74       22.78   
78 2014-11-04  United States Senate    Crockett       67.97       25.30   
79 2014-11-04  United States Senate      Coffee       65.47       26.82   
80 2014-11-04  United States Senate       Cocke       60.08       34.19   
81 2014-11-04  United States Senate        Clay       61.61       31.86   
82 2014-11-04  United States Senate   Claiborne       69.32       24.88   
83 2014-11-04  United States Senate     Chester       70.74       18.98   
84 2014-11-04  United States Senate    Cheatham       64.32       27.46   
85 2014-11-04  United States Senate      Carter       73.54       18.91   
86 2014-11-04  United States Senate     Carroll       65.61       26.90   
87 2014-11-04  United States Senate      Cannon       62.36       29.08   
88 2014-11-04  United States Senate    Campbell       70.46       24.14   
89 2014-11-04  United States Senate     Bradley       78.87       16.02   
90 2014-11-04  United States Senate      Blount       68.84       23.97   
91 2014-11-04  United States Senate     Bledsoe       69.45       24.34   
92 2014-11-04  United States Senate      Benton       58.89       34.58   
93 2014-11-04  United States Senate     Bedford       66.68       26.37   
94 2014-11-04  United States Senate    Anderson       64.53       29.24   

    Other  Rep_Margin  
70  12.48       46.20  
71   4.94       46.38  
72   6.14       48.04  
73   8.51       30.30  
74   7.61       26.82  
75   5.36       33.06  
76   5.95       -7.07  
77   6.41       46.96  
78   4.02       42.67  
79   7.18       38.65  
80   6.86       25.89  
81   7.17       29.75  
82   6.85       44.44  
83   7.10       51.76  
84   8.35       36.86  
85   8.01       54.63  
86   6.49       38.71  
87   7.50       33.28  
88   6.74       46.32  
89   7.64       62.85  
90   6.52       44.87  
91   7.77       45.11  
92   9.47       24.31  
93   6.53       40.31  
94   7.26       35.29

In [55]:
tn_2014_uss_county_pct.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 95 entries, 0 to 94
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   electdate   95 non-null     datetime64[ns]
 1   officename  95 non-null     object        
 2   county      95 non-null     object        
 3   Republican  95 non-null     float64       
 4   Democratic  95 non-null     float64       
 5   Other       95 non-null     float64       
 6   Rep_Margin  95 non-null     float64       
dtypes: datetime64[ns](1), float64(4), object(2)
memory usage: 5.3+ KB


In [56]:
tn_2014_uss_county_pct.to_excel(r'../data/2014_election/tn_2014_uss_county_pct.xlsx',index=False)